### ISSUE TITLE
The vector.push_back handling

### ISSUE THREAD

**Issue by qian-feng** [seahorn/seahorn#148]
I am trying to utilize seahorn to check the memory safety of a small example. It seems that seahorn runs forever for that example.  The example is as following:

 #include <iostream>
#include <string>
#include <vector>
#include "seahorn/seahorn.h"

using namespace std;
extern int nd();
vector<string> v_str;

string test6(int index) {
  v_str.push_back("xxx");
  v_str.push_back("yyy");
  v_str.push_back("zzz");
  v_str.push_back("test6");
  return v_str[index];
}

int main(int argc, char ** argv) {
  int index = nd();
  cout << test6(index) << "\n";
  return 0;
}

Does it mean seahorn do not support the memory handling for vector in C++, especially for the push_back operation? Thanks.

Best,

**Comment by agurfinkel** [comment]
By default, `SeaHorn` only checks for assertions preset in the code. We are currently working on automatic synthesis of assertions for memory safety, but this is not yet ready for prime time. In the past, we have created memory safety checkers targeting C code, but those tend not to work very well with C++ code bases. What options are you using to run seahorn on your example?

**Comment by qian-feng** [comment]
i used "sea abc -O0 test.cpp --show-invars".
Has seahorn already support the push_back operation in Vector? I tried the simple c++ example, it seams the assertion result is not correct.
#include <iostream>
#include <vector>
extern void __VERIFIER_assume (int);
extern void __VERIFIER_error (void);
__attribute__((always_inline)) void assert  (int v)  { if (!v) __VERIFIER_error (); }
int main ()
{
  std::vector<int> fifth;                       // a copy of third
  int i=0;

  // the iterator constructor can also be used to construct from arrays:
  for(i=0; i<5; i++)
    fifth[i]=1;

  assert(fifth[0]>5);

  return 0;
}
Thanks. I really appreciate your help.

Best,

**Comment by agurfinkel** [comment]
`sea abc` is an experimental option that is currently under development. It is not guaranteed to work outside of the examples we are developing it right now. We are very interested in applications, especially memory safety. We would appreciate any examples / benchmarks / problems that you can share with us.

`SeaHorn` by itself does not provide support for any external libraries. To "teach" it about vector you need to include the source code of `std::vector` as part of the translation unit that it given to `seahorn`. Otherwise, the calls to the `std::vector` library appear external and are treated as non-deterministic.

Including an efficient implementation of STL data types is not likely to scale. Alternatively, you can develop verification stubs that expose the desired semantics (for example, implementing a vector by C array). The kind of a stub will depend on what you are trying to verify and what is assumed about the library.

We develop such libraries per each project that we verify. However, we do not have publicly available examples of such libraries. 

In summary, SeaHorn only knows about LLVM bitcode. It can verify programs in LLVM bitcode by treating all external code as non-deterministic. If you want to use SeaHorn with third-party libraries, you have to either provide code for them or verification stubs.
